File: Data Structures & Algorithms/invert-a-binary-tree/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func invertTree(root *TreeNode) *TreeNode {
	if root == nil{
		return root
	}

	left := invertTree(root.Right)
	right := invertTree(root.Left)

	root.Left = left
	root.Right = right

	return root

}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void dfs(TreeNode* root,int m,int &count) // here if we pass the address of m, when our m
    //gets updated with large value at the end and be backtrack, the large value remains.
    // if we pass just m, then we retain a copy of m at every recursive call, so value of m in call not changed by other recursive call
    {
        if(!root)
        {
            return;
        }
        if(root->val >= m)
        {
            m = root->val;
            count++;
        }

        dfs(root->left,m,count);
        dfs(root->right,m,count);

        return;
    }
    int goodNodes(TreeNode* root) {
        int m=INT_MIN;
        int count = 0;
        dfs(root,m,count);
        return count;
        
    }
};
